File: server/proto/bridge.go
package proto

import "time"

// The durable outcome of the last apply, read back by a client whose connection
// the apply itself cut: enabling the bridge moves the management address, so
// losing the response is the expected case and GET is how it learns what
// happened.
type BridgeState string

const (
	// The stock device: no br0, the address on eth0, no l2-uplink file. A device
	// that never bridged and one that was disabled are byte-identical here.
	BridgeDisabled BridgeState = "disabled"

	BridgeEnabled BridgeState = "enabled"

	// A gate failed or the deadline expired and the snapshot was restored.
	BridgeRolledBack BridgeState = "rolledBack"

	// The restore itself did not complete. The device may need the wlan0 AP or
	// a serial console.
	BridgeFailed BridgeState = "failed"

	// A pending.json is armed: an apply is in flight, or the device booted into
	// an unfinished one.
	BridgePending BridgeState = "pending"
)

// All three gates must be true before the dead-man is disarmed. Address and
// Gateway together are only L3 reachability; Inbound is what proves the
// management plane answers, which is a different property on a device whose
// S95nanokvm installs DROP OUTPUT tcp --sport 8000.
type BridgeChecks struct {
	Address bool `json:"address"`
	Gateway bool `json:"gateway"`
	Inbound bool `json:"inbound"`

	// Inbound was satisfied by the self-connect fallback rather than a real
	// client, which proves the listener and local delivery, not the wire.
	InboundWeak bool `json:"inboundWeak"`
}

func (c BridgeChecks) Passed() bool {
	return c.Address && c.Gateway && c.Inbound
}

type BridgePort struct {
	Name  string `json:"name"`
	State string `json:"state"`
	Up    bool   `json:"up"`

	// Carrier is IFF_LOWER_UP: the cable, as opposed to Up's administrative
	// state, which is set on a port with nothing plugged into it too.
	Carrier bool `json:"carrier"`
}

// A second path between the uplink's segment and a port of the same bridge.
// STP is off by design, because its listening delay costs the DHCP lease, so
// nothing in the kernel breaks such a loop and this warning is the compensating
// control. Non-nil is evidence the condition exists; nil is not evidence that
// it does not, since the entry it is read from moves between ports.
type BridgeLoop struct {
	Port   string `json:"port"`
	MAC    string `json:"mac"`
	Reason string `json:"reason"`
}

type BridgeArmed struct {
	Operation    string    `json:"operation"`
	SnapshotPath string    `json:"snapshotPath"`
	ArmedAt      time.Time `json:"armedAt"`
	Deadline     time.Time `json:"deadline"`
}

type BridgeApply struct {
	State     BridgeState  `json:"state"`
	Uplink    string       `json:"uplink"`
	Enabled   bool         `json:"enabled"`
	Checks    BridgeChecks `json:"checks"`
	Message   string       `json:"message"`
	AppliedAt time.Time    `json:"appliedAt"`
}

type GetBridgeRsp struct {
	State   BridgeState  `json:"state"`
	Uplink  string       `json:"uplink"`
	Exists  bool         `json:"exists"`
	MAC     string       `json:"mac"`
	Ports   []BridgePort `json:"ports"`
	Address string       `json:"address"`
	Gateway string       `json:"gateway"`

	// Carrier on the device Uplink names, which is br0 once the bridge holds
	// the address and eth0 otherwise. A false here with the bridge enabled is
	// the state the enable preflight exists to refuse to create.
	Carrier bool `json:"carrier"`

	Loop      *BridgeLoop  `json:"loop"`
	Pending   *BridgeArmed `json:"pending"`
	LastApply *BridgeApply `json:"lastApply"`

	// ncm, rndis, or empty for a gadget presenting no network. Reported from
	// the presentation snapshot, because the protocol the gadget presents is a
	// property of the USB profile and not of the bridge. The control for it
	// lives on the Virtual Network switch under Settings, Device.
	Protocol string `json:"protocol"`
}

type SetBridgeReq struct {
	Enabled bool `json:"enabled"`
}

type SetBridgeRsp struct {
	State   BridgeState  `json:"state"`
	Uplink  string       `json:"uplink"`
	Checks  BridgeChecks `json:"checks"`
	Message string       `json:"message"`
}

File: web/src/api/bridge.ts
import { http } from '@/lib/http.ts';

// the transaction runs inside the request and holds the management address for
// the whole verification window, so it outlives the client's default timeout
const applyTimeout = 90 * 1000;

export type BridgeState = 'disabled' | 'enabled' | 'rolledBack' | 'failed' | 'pending';

// the three verification gates, plus a note that inbound was only self-tested
export type BridgeChecks = {
  address: boolean;
  gateway: boolean;
  inbound: boolean;
  inboundWeak: boolean;
};

export type BridgePort = {
  name: string;
  state: string;
  up: boolean;
  // the cable, as opposed to up, which is set on a port with nothing in it too
  carrier: boolean;
};

// a second path between the uplink's segment and another port of the bridge.
// Non-null is evidence the condition exists; null is not evidence it does not.
export type BridgeLoop = {
  port: string;
  mac: string;
  reason: string;
};

// an armed dead-man marker: an apply is in flight, or the device booted into
// one that never finished
export type BridgeArmed = {
  operation: string;
  snapshotPath: string;
  armedAt: string;
  deadline: string;
};

export type BridgeApply = {
  state: BridgeState;
  uplink: string;
  enabled: boolean;
  checks: BridgeChecks;
  message: string;
  appliedAt: string;
};

export type BridgeStatus = {
  state: BridgeState;
  uplink: string;
  exists: boolean;
  mac: string;
  ports: BridgePort[];
  address: string;
  gateway: string;
  carrier: boolean;
  loop?: BridgeLoop | null;
  pending?: BridgeArmed | null;
  lastApply?: BridgeApply | null;

  // ncm, rndis, or empty for a gadget presenting no network. Reported only: the
  // control for it is the Virtual Network one under Settings, Device, because
  // the protocol is a property of the USB profile rather than of the bridge.
  protocol: string;
};

// get the bridge state, its ports, the armed marker and the last transaction
export function getBridge() {
  return http.get('/api/network/bridge');
}

// enable or disable the bridge; the reply repeats the state, the uplink, the
// gates and a message, but the refetch is what a caller who lost it reads back
export function setBridge(enabled: boolean) {
  return http.post('/api/network/bridge', { enabled }, { timeout: applyTimeout });
}

// put the snapshot back after an apply that never finished
export function revertBridge() {
  return http.post('/api/network/bridge/revert', {}, { timeout: applyTimeout });
}
